File: apps/api/replaytutor/routes/chart_tools.py
from fastapi import APIRouter, HTTPException, Request, Response, status

from replaytutor.config import Settings
from replaytutor.contracts import (
    ChartTool,
    ChartToolManifestListResponse,
    ChartToolPreference,
    ChartToolPreferenceListResponse,
    ChartToolTemplate,
    ChartToolTemplateListResponse,
    CreateChartToolTemplateRequest,
    UpdateChartToolPreferenceRequest,
)
from replaytutor.modules.chart_tools import ChartToolService

router = APIRouter(prefix="/api/v1/chart-tools", tags=["chart-tools"])


def service(request: Request) -> ChartToolService:
    settings: Settings = request.app.state.settings
    return ChartToolService(settings)


@router.get("", response_model=ChartToolManifestListResponse)
def list_chart_tools(request: Request) -> ChartToolManifestListResponse:
    return service(request).manifests()


@router.get("/templates", response_model=ChartToolTemplateListResponse)
def list_templates(request: Request) -> ChartToolTemplateListResponse:
    return service(request).list_templates()


@router.post("/templates", response_model=ChartToolTemplate)
def create_template(
    request: Request,
    payload: CreateChartToolTemplateRequest,
) -> ChartToolTemplate:
    return service(request).create_template(payload)


@router.delete("/templates/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_template(request: Request, template_id: str) -> Response:
    try:
        service(request).delete_template(template_id)
    except ValueError as error:
        raise HTTPException(status_code=404, detail=str(error)) from error
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/preferences", response_model=ChartToolPreferenceListResponse)
def list_preferences(request: Request) -> ChartToolPreferenceListResponse:
    return service(request).list_preferences()


@router.put("/preferences/{tool}", response_model=ChartToolPreference)
def update_preference(
    request: Request,
    tool: ChartTool,
    payload: UpdateChartToolPreferenceRequest,
) -> ChartToolPreference:
    try:
        return service(request).update_preference(tool, payload)
    except ValueError as error:
        raise HTTPException(status_code=422, detail=str(error)) from error

File: apps/web/src/api/chartTools.ts
import type {
  ChartAnnotation,
  ChartToolManifestListResponse,
  ChartToolPreference,
  ChartToolPreferenceListResponse,
  ChartToolTemplate,
  ChartToolTemplateListResponse,
  CreateChartToolTemplateRequest,
  UpdateChartToolPreferenceRequest,
} from "@replaytutor/contracts";
import { validateContract } from "@replaytutor/contracts";
import { API_BASE_URL, ApiResponseError, apiFetch } from "./health";

async function contractJson<T>(response: Response, contract: Parameters<typeof validateContract>[0]): Promise<T> {
  if (!response.ok) throw new ApiResponseError(`Chart tool request failed: ${response.status}`, response.status);
  const value: unknown = await response.json();
  if (!validateContract<T>(contract, value)) throw new ApiResponseError(`Invalid ${contract} response`, 502);
  return value;
}

export function fetchChartToolManifest(): Promise<ChartToolManifestListResponse> {
  return apiFetch(`${API_BASE_URL}/api/v1/chart-tools`).then((response) => contractJson(response, "ChartToolManifestListResponse"));
}

export function fetchChartToolTemplates(): Promise<ChartToolTemplateListResponse> {
  return apiFetch(`${API_BASE_URL}/api/v1/chart-tools/templates`).then((response) => contractJson(response, "ChartToolTemplateListResponse"));
}

export function createChartToolTemplate(payload: CreateChartToolTemplateRequest): Promise<ChartToolTemplate> {
  return apiFetch(`${API_BASE_URL}/api/v1/chart-tools/templates`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  }).then((response) => contractJson(response, "ChartToolTemplate"));
}

export function fetchChartToolPreferences(): Promise<ChartToolPreferenceListResponse> {
  return apiFetch(`${API_BASE_URL}/api/v1/chart-tools/preferences`).then((response) => contractJson(response, "ChartToolPreferenceListResponse"));
}

export function updateChartToolPreference(
  tool: NonNullable<ChartAnnotation["tool"]>,
  payload: UpdateChartToolPreferenceRequest,
): Promise<ChartToolPreference> {
  return apiFetch(`${API_BASE_URL}/api/v1/chart-tools/preferences/${encodeURIComponent(tool)}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  }).then((response) => contractJson(response, "ChartToolPreference"));
}
